logstash-core/src/main/java/org/logstash/ackedqueue/SettingsImpl.java

```diff
@@ -0,0 +1,252 @@
+        this.dirForFiles = dirForFiles;
+        this.checkpointIOFactory = checkpointIOFactory;
+        this.pageIOFactory = pageIOFactory;
+        this.elementClass = elementClass;
+        this.capacity = capacity;
+        this.queueMaxBytes = queueMaxBytes;
+        this.maxUnread = maxUnread;
+        this.checkpointMaxAcks = checkpointMaxAcks;
+        this.checkpointMaxWrites = checkpointMaxWrites;
+    }
+
+    @Override
+    public int getCheckpointMaxAcks() {
+        return checkpointMaxAcks;
+    }
+
+    @Override
+    public int getCheckpointMaxWrites() {
+        return checkpointMaxWrites;
+    }
+    
+    @Override
+    public CheckpointIOFactory getCheckpointIOFactory() {
+        return checkpointIOFactory;
+    }
+
+    public PageIOFactory getPageIOFactory() {
+        return pageIOFactory;
+    }
+
+    @Override
+    public Class getElementClass()  {
+        return this.elementClass;
+    }
+
+    @Override
+    public String getDirPath() {
+        return dirForFiles;
+    }
+
+    @Override
+    public long getQueueMaxBytes() {
+        return queueMaxBytes;
+    }
+
+    @Override
+    public int getCapacity() {
+        return capacity;
+    }
+
+    @Override
+    public int getMaxUnread() {
+        return this.maxUnread;
+    }
+
+    private static final class BuilderImpl implements Builder {
+        
+        /**
+         * The default Queue has a capacity of 0 events, meaning infinite capacity.
+         * todo: Remove the ability to set infinite capacity.
+         */
+        private static final int DEFAULT_CAPACITY = 0;
+
+        /**
+         * The default Queue has a capacity of 0 bytes, meaning infinite capacity.
+         * todo: Remove the ability to set infinite capacity.
+         */
+        private static final long DEFAULT_MAX_QUEUE_BYTES = 0L;
+
+        /**
+         * The default max unread count 0, meaning infinite.
+         * todo: Remove the ability to set infinite capacity.
+         */
+        private static final int DEFAULT_MAX_UNREAD = 0;
+
+        /**
+         * Default max number of writes after which we checkpoint.
+         */
+        private static final int DEFAULT_CHECKPOINT_MAX_ACKS = 1024;
+
+        /**
+         * Default number of acknowledgements after which we checkpoint.
+         */
+        private static final int DEFAULT_CHECKPOINT_MAX_WRITES = 1024;
+        
+        private final String dirForFiles;
+
+        private final CheckpointIOFactory checkpointIOFactory;
+
+        private final PageIOFactory pageIOFactory;
+
+        private final Class elementClass;
+
+        private final int capacity;
+
+        private final long queueMaxBytes;
+
+        private final int maxUnread;
+
+        private final int checkpointMaxAcks;
+
+        private final int checkpointMaxWrites;
+        
+        private BuilderImpl(final String dirForFiles) {
+            this(dirForFiles, null, null, null, DEFAULT_CAPACITY, DEFAULT_MAX_QUEUE_BYTES,
```

Comment:
@robbavey I think these `null` values can also be killed easily. It seems to me there is a clear set of combinations we can have here, that we can probably leverage to fix this clean. (just adding the note so I don't forget to add an issue)